File: BOJ/5972.cpp
#include<bits/stdc++.h>
#define INF 987654321
using namespace std;

int dist[50001]; // 시작 node로 부터 i번째 노드까지의 최단 거리
bool visited[50001];
priority_queue<pair<int, int>, vector<pair<int, int> >, greater<pair<int, int> > > pq; // 비용이 작은 경로를 우선 탐색하기 위한 우선순위 큐
vector<pair<int, int> > vp[50001]; // 노드의 연결을 표현하기 위함 (연결된 노드, 비용) 쌍

void dijkstra(int start, int end) {
	pq.push(make_pair(0, start)); // 시작 지점의 비용(거리), 노드를 저장 해준다.
	dist[start] = 0; // 시작 지점으로부터의 거리는 당연히 0이다.

	while (!pq.empty()) {
		int cost = pq.top().first;
		int cur_node = pq.top().second;
		pq.pop();

		if (cur_node == end) return; // 목표 지점에 도착 했으면 더 이상 진행하지 않는다.
		if (visited[cur_node]) continue; // 이미 방문한 노드이면 pass한다.

		visited[cur_node] = true;
		for (int i = 0; i < vp[cur_node].size(); i++) {
			int next = vp[cur_node][i].first;
			int n_cost = vp[cur_node][i].second;

			// 방문할 노드의 거리가 (현재까지의 cost + 다음 노드로 이동할 때의 cost)보다 크다면 새로 갱신해주며 이동.
			if (dist[next] > cost + n_cost) {
				dist[next] = cost + n_cost;
				pq.push(make_pair(dist[next], next));
			}
		}
	}
}

int main() {
	int N, M;
	cin >> N >> M;
	for (int i = 0; i < M; i++) {
		int u, v, cost;
		cin >> u >> v >> cost;
		vp[u].push_back(make_pair(v, cost)); // 양 방향으로 노드를 연결해준다.
		vp[v].push_back(make_pair(u, cost));
	}

	for (int i = 1; i <= N; i++) dist[i] = INF; // 나올 수 없는 큰 값을 비용에 넣어준다. (최소한의 비용으로 갱신하기 위함)

	dijkstra(1, N);
	cout << dist[N] << '\n';

	return 0;
}

File: BOJ/python/5972.py
import sys
import heapq as hq  # 이번엔 heapq를 써서 해보았다. PriorityQueue랑 똑같다.

input = sys.stdin.readline
INF = sys.maxsize

N, M = map(int, input().rstrip("\n").split(" "))
vector = [[] for _ in range(N + 1)]
visited = [False for _ in range(N + 1)]
dist = [INF] * (N + 1)

for _ in range(M):
    u, v, cost = map(int, input().rstrip("\n").split(" "))
    vector[u].append((v, cost))  # 방향성 없는 그래프이기 때문에 서로 연결해준다.
    vector[v].append((u, cost))


def dijkstra(start, end, vector: list) -> None:
    heap_min = []
    hq.heappush(heap_min, (0, start))
    dist[start] = 0

    while heap_min:
        cost, cur_node = hq.heappop(heap_min)

        if visited[cur_node]:  # 이미 방문한 노드이면 최소 비용이 담겨 있기 때문에 무시한다.
            continue

        visited[cur_node] = True
        for next_node, next_cost in vector[cur_node]:

            # 다음 노드까지의 비용이 (현재까지의 비용 + 다음 노드로 이동하는 비용) 보다 크다면 새로 갱신해준다.
            if dist[next_node] > cost + next_cost:
                dist[next_node] = cost + next_cost
                hq.heappush(heap_min, (dist[next_node], next_node))


dijkstra(1, N, vector)
print(dist[N])
